### PR TITLE
Change max shingle size and the number of trees

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are inter-file references, I split a big PR into multiple smaller PRs based on files to save time (1.1 cut off is due at the end of this week). The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big PR in the end and merge once after all review PRs get approved. For the complete code, check https://github.com/kaituo/anomaly-detection-1/tree/compactRCF2

### Description
The PR made the following changes.

1)Change to limit max shingle size to 1024. Previously, it was unbounded. The reasons are twofold. First, it may not work on the RCF side. It is unknown that a huge shingle size could give better accuracy. Second, the larger the shingle, the more resources (e.g., CPU for tree construction/traversal, memory for models, and disk for checkpoint) RCF will use. For example, a compact RCF model with 100 trees and 932 dimensions uses around 241 MB of memory. The same model in RCF 1.0 uses 958 MB.
2)Change the number of trees used in HCAD RCF from 10 to 30 trees. Ten trees are too aggressive and may hurt precision.

Testing done:
1. tested the change does not impact existing functionality.
 
### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/97
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java [hunk 1]
```diff
@@ -644,6 +651,25 @@ private AnomalyDetectorSettings() {}
...
+    // ======================================
+    // Checkpoint setting
+    // ======================================
+    // we won't accept a checkpoint larger than 30MB. Or we risk OOM.
+    // For reference, in RCF 1.0, the checkpoint of a RCF with 50 trees, 10 dimensions,
+    // 256 samples is of 3.2MB.
+    // In compact rcf, the same RCF is of 163KB.
+    // Since we allow at most 5 features, and the default shingle size is 8 and default
+    // tree number size is 100, we can have at most 25.6 MB. It is possible that cx increases
+    // the max features or shingle size, but we don't want to risk OOM for the flexibility.
```
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:663]: Any experiments done on increasing these defaults with compact RCF, like higher default tree size or shingle size?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:663]: There are some experiments here done by Sudipto/Josh: https://quip-amazon.com/A8hTAbgyHFtF/Random-Cut-Forest-serialization-benchmarks

Comment (ohltyler) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:663]: Ok, will take a look. As a reminder, should not have inaccessible links in public PRs 😃

**Review by ohltyler — DISMISSED** [review]

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java [hunk 2]
```diff
@@ -211,8 +212,13 @@ public AnomalyDetector(
         if (detectionInterval == null) {
             throw new IllegalArgumentException("Detection interval should be set");
         }
-        if (shingleSize != null && shingleSize < 1) {
-            throw new IllegalArgumentException("Shingle size must be a positive integer");
+        if (shingleSize != null && (shingleSize < 1 || shingleSize > AnomalyDetectorSettings.MAX_SHINGLE_SIZE)) {
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/model/AnomalyDetector.java:215]: Thanks for adding this check. How about build a util method so historical analysis can reuse it?

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java [hunk 3]
```diff
@@ -222,6 +222,21 @@ private AnomalyDetectorSettings() {}
             Setting.Property.Dynamic
         );
 
+    // for a batch operation, we want all of the bounding box in-place for speed
+    public static final double BATCH_BOUNDING_BOX_CACHE_RATIO = 1;
+
+    // for a real-time operation, we trade off speed for memory as real time opearation
+    // only has to do one update/scoring per interval
+    public static final double REAL_TIME_BOUNDING_BOX_CACHE_RATIO = 0;
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:230]: Does this mean we need to rebuild bounding box for every realtime job run? How much more memory will cost if we open this bounding box cache per RCF model?

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java [hunk 4]
```diff
@@ -222,6 +222,21 @@ private AnomalyDetectorSettings() {}
...
+
+    public static final int DEFAULT_SHINGLE_SIZE = 8;
+
+    // Multi-entity detector model setting:
+    // TODO (kaituo): change to 4
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:235]: This TODO means we need to change this hard coded HCAD shingle size from 1 to 4? Or we should set the HCAD shingle size upper limit as 4 (or some other value)?

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:215]: sure, added

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:230]: Different config have different numbers. For 30 trees, 2 dimensions, using full bounding box costs 935 KB, while not using bounding box  costs 322 KB.

Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:230]: Yes, it means we need to rebuild for every real time job run.

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java (hunk 4 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:235]: It means we should have default value 4 but allow people to change like single-stream detectors.

Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:235]: Got it. Should we set different upper limit of shingle size for single-stream detector and HCAD?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:235]: We can do that when there is a need.

Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:235]: sure

File: src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:663]: right, next time will send you offline.

**Review by sudiptoguha — DISMISSED** [review]
ylwu-amzn should eyeball changes too. Looks good to me. Question: is the historical analysis using 100 trees? Did anyone try the getApproximateScore()? That may speed up considerably for larger number of trees.

File: src/main/java/org/opensearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:215]: Not sure if I missed something, but seems the util method not added yet?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/AnomalyDetector.java:215]: not yet.  Will send the new commit after retesting.

**Comment by kaituo** [comment]
> ylwu-amzn should eyeball changes too. Looks good to me. Question: is the historical analysis using 100 trees? Did anyone try the getApproximateScore()? That may speed up considerably for larger number of trees.

for HCAD, a historical detector uses 30 trees; for single stream  a historical detector uses 100 trees. As we discussed offline, using getApproximateScore for single stream complicates the logic we have multiple-way split and run distributed models.

**Comment by ylwu-amzn** [comment]
> > ylwu-amzn should eyeball changes too. Looks good to me. Question: is the historical analysis using 100 trees? Did anyone try the getApproximateScore()? That may speed up considerably for larger number of trees.
> 
> for HCAD, a historical detector uses 30 trees; for single stream a historical detector uses 100 trees. As we discussed offline, using getApproximateScore for single stream complicates the logic as we have multiple-way split and run distributed models.

Not sure if understand it correctly. I think `getApproximateScore` will run faster but it may get approximate score. @sudiptoguha, can you confirm if that approximate score will be highly confident and won't impact the detection result? 
For historical analysis, we have add rate limiting, we run 1000 detection interval in one batch, then run the next batch in 5 seconds. So the speed is not top bottleneck for historical analysis considering the batch interval time(5 seconds) is dominating the speed. But that may help for HC historical analysis if we multiply by huge amount of entities.

**Comment by sudiptoguha** [comment]
> > > ylwu-amzn should eyeball changes too. Looks good to me. Question: is the historical analysis using 100 trees? Did anyone try the getApproximateScore()? That may speed up considerably for larger number of trees.
> > 
> > 
> > for HCAD, a historical detector uses 30 trees; for single stream a historical detector uses 100 trees. As we discussed offline, using getApproximateScore for single stream complicates the logic as we have multiple-way split and run distributed models.
> 
> Not sure if understand it correctly. I think `getApproximateScore` will run faster but it may get approximate score. @sudiptoguha, can you confirm if that approximate score will be highly confident and won't impact the detection result?
> For historical analysis, we have add rate limiting, we run 1000 detection intervals in one batch, then run the next batch in 5 seconds. So the speed is not top bottleneck for historical analysis considering the batch interval time(5 seconds) is dominating the speed. But that may help for HC historical analysis if we multiply by huge amount of entities.

When the model has enough samples (not the cold start scenario) then the approximateScore() will be fine -- those are the tests I saw. Not sure about both cold start (low data) and approximation ... I think it may be a good call to enable approximation for the "second 1000 intervals onwards" -- what do you think? Because when we have 1000+ intervals we must have higher threshold to not have too many anomaly alerts, and in that case we can use approximation. Also I would not use approximation for 30 trees (its already small) -- for 100 trees, the benefit kicks in because we would use 30ish trees in the average case (to decide the easy "not an anomaly" and easy "anomaly" cases) -- we may use the full 100 to decide the ambiguous cases.

**Comment by kaituo** [comment]
> > > > ylwu-amzn should eyeball changes too. Looks good to me. Question: is the historical analysis using 100 trees? Did anyone try the getApproximateScore()? That may speed up considerably for larger number of trees.
> > > 
> > > 
> > > for HCAD, a historical detector uses 30 trees; for single stream a historical detector uses 100 trees. As we discussed offline, using getApproximateScore for single stream complicates the logic as we have multiple-way split and run distributed models.
> > 
> > 
> > Not sure if understand it correctly. I think `getApproximateScore` will run faster but it may get approximate score. @sudiptoguha, can you confirm if that approximate score will be highly confident and won't impact the detection result?
> > For historical analysis, we have add rate limiting, we run 1000 detection intervals in one batch, then run the next batch in 5 seconds. So the speed is not top bottleneck for historical analysis considering the batch interval time(5 seconds) is dominating the speed. But that may help for HC historical analysis if we multiply by huge amount of entities.
> 
> When the model has enough samples (not the cold start scenario) then the approximateScore() will be fine -- those are the tests I saw. Not sure about both cold start (low data) and approximation ... I think it may be a good call to enable approximation for the "second 1000 intervals onwards" -- what do you think? Because when we have 1000+ intervals we must have higher threshold to not have too many anomaly alerts, and in that case we can use approximation. Also I would not use approximation for 30 trees (its already small) -- for 100 trees, the benefit kicks in because we would use 30ish trees in the average case (to decide the easy "not an anomaly" and easy "anomaly" cases) -- we may use the full 100 to decide the ambiguous cases.

@sudiptoguha  I changed single stream detectors' trees to 30 according to your suggestion.  So all RCF (including historical and single-stream) models use 30 trees inside AD.

**Review by sudiptoguha — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Pull request closed without merging**